### ISSUE TITLE
looks like `model.enable_input_require_grads()` only enable if `training_args.use_lora` is True?

### ISSUE THREAD

**Issue by tiendung** [QwenLM/Qwen#405]
https://github.com/QwenLM/Qwen/blob/f0078d013b3fe3de22b83088c8b17254b40440f8/finetune.py#L342

**Comment by JustinLin610** [comment]
Oh, I am going to add an explanation for this. This results from the chatml format. If you train the model from the base model, e.g., Qwen-7B, instead of chat models, e.g., Qwen-7B-Chat, you should enable the training of the embedding and output layer for the model to learn to predict new tokens, which are special tokens. However, if you use Q-LoRA, you cannot train the embeddings. Since the int4 models that we provide are chat models, then you do not have such worries.

**Comment by xwk** [comment]
@JustinLin610 Is thes 2 lines correct? 

https://github.com/QwenLM/Qwen/blob/3ad0c83bb9a31fb3ccb8b70b6ec5a92caa4f7170/finetune.py#L354-L355

From your description, model.enable_input_require_grads() should be called when add new special token to the vocabulary, but what lines check is whether the gradient_checkpointing is enabled, which seems to be totally irrelevant.

**Comment by redagavin** [comment]
> @JustinLin610 are these 2 lines correct?
> 
> https://github.com/QwenLM/Qwen/blob/3ad0c83bb9a31fb3ccb8b70b6ec5a92caa4f7170/finetune.py#L354-L355
> 
> From your description, model.enable_input_require_grads() should be called when add new special token to the vocabulary, but what these lines check is whether the gradient_checkpointing is enabled, which seems to be totally irrelevant.

I'm also confused by these two lines. @jklj077 @JustinLin610 Could you please explain why training input embedding is enabled when gradient checkpointing is used?

**Comment by btcnhung1299** [comment]
I would also like to know why model.enable_input_require_grads() is called for gradient checkpointing.
